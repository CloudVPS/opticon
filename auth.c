#include <auth.h>

// http://www.literatecode.com/get/aes256.c

/** Initialize session-related global storage. */
void sessionlist_init (void) {
	for (int i=0; i<256; ++i) {
		SESSIONS[i].first = SESSIONS[i].last = NULL;
	}
}

/** Allocate and initialize session object. */
session *session_alloc (void) {
	session *res = (session *) malloc (sizeof (session));
	res->next = res->prev = NULL;
	res->addr = res->sessid = 0;
	res->key.data = {0,0,0,0,0,0,0,0};
	res->lastcycle = time (NULL);
	res->host = NULL;
	return res;
}

/** Link a new session into the proper bucket. */
void session_link (session *s) {
	uint32_t sid = s->sessid ^ s->addr;
	uint8_t bucket = (sid >> 24) |
					 ((sid & 0x00ff0000) >> 16) |
					 ((sid & 0x0000ff00) >> 8) |
					 (sid & 0xff);
	if (SESSIONS[bucket].first) {
		// link in at top.
		s->next = SESSIONS[bucket].first;
		SESSIONS[bucket].first->prev = s;
		SESSIONS[bucket].first = s;
	}
	else {
		SESSIONS[bucket].first = SESSIONS[bucket].last = s;
	}
}

/** Register a new session (or update an existing session to
  * stop it from expiring. */
int session_register (uuid tenantid, uuid hostid,
					  uint32_t addrpart, uint32_t sess_id.
					  aeskey sess_key) {
	host *h = host_find (tenantid, hostid);
	session *s;
	if (! h) return 0;

	if (s = session_find (hostid, addrpart)) {
		s->lastcycle = time (NULL);
		return 1;
	}
		
	s = session_alloc();
	if (! s) return 0;
	
	s->tenantid = tenantid;
	s->hostid = hostid;
	s->addr = addrpart;
	s->sessid = sess_id;
	s->key = sess_key;
	
	session_link (s);
	return 1;
}

session *session_find (uint32_t addrpart, uint32_t sess_id) {
	uint32_t sid = s->sessid ^ s->addr;
	uint8_t bucket = (sid >> 24) |
					 ((sid & 0x00ff0000) >> 16) |
					 ((sid & 0x0000ff00) >> 8) |
					 (sid & 0xff);
	session *s = SESSIONS[bucket].first;
	while (s) {
		if (s->addr == addrpart && s->sessid == sess_id) return s;
		s = s->next;
	}
	return NULL;
}

void session_expire (time_t cutoff) {
	session *s;
	for (uint8_t i=0; i<=255; ++i) {
		s = SESSIONS[i].first;
		while (s) {
			if (s->lastcycle < cutoff) {
				if (s->prev) {
					if (s->next) {
						s->prev->next = s->next;
						s->next->prev = s->prev;
					}
					else {
						s->prev->next = NULL;
						SESSIONS[i].last = s->prev;
					}
				}
				else {
					if (s->next) {
						s->next->prev = NULL;
						SESSIONS[i].first = s->next;
					}
					else {
						SESSIONS[i].first = SESSIONS[i].last = NULL;
					}
				}
				
				free (s);
			}
			s = s->next;
		}
	}
}
