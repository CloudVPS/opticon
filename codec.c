#include <codec.h>
#include <stdlib.h>

codec *codec_create_json (void) {
    codec *res = (codec *) malloc (sizeof (codec));
    if (! res) return res;
    res->encode_host = jsoncodec_encode_host;
    res->decode_host = jsoncodec_decode_host;
    return res;
}

codec *codec_create_opticon (void) {
    return NULL;
}

void codec_release (codec *c) {
    free (c);
}

int codec_encode_host (codec *c, ioport *io, host *h) {
    return c->encode_host (io, h);
}

int codec_decode_host (codec *c, ioport *io, host *h) {
    return c->decode_host (io, h);
}

int jsoncodec_encode_host (ioport *io, host *h) {
    dump_host_json (h, io);
    return 1;
}

int jsoncodec_decode_host (ioport *io, host *h) {
    return 0;
}
