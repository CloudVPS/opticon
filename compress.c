#include <compress.h>

int compress_data (ioport *in, ioport *out) {
    if (ioport_read_available (in) == 0) return 0;
    
    z_stream strm;
    strm.zalloc = Z_NULL;
    strm.zfree = Z_NULL;
    strm.opaque = Z_NULL;
    strm.total_out = 0;
    strm.next_in = (Bytef*)ioport_get_buffer (in);
    strm.avail_in = ioport_read_available (in);
    
    int r = deflateInit2 (&strm, Z_DEFAULT_COMPRESSION,
                          Z_DEFLATED, 15+16, 8,
                          Z_DEFAULT_STRATEGY);
    if (r != Z_OK) return 0;
    
    int st;
    do {
        strm.next_out = (Bytef*)ioport_get_buffer (out) + strm.total_out;
        strm.avail_out = ioport_write_available (out) - strm.total_out;
        st = deflate (&strm, Z_FINISH);
    } while (st == Z_OK);
    
    if (st != Z_STREAM_END) return 0;
    ioport_write (out, ioport_get_buffer (out), strm.total_out);
    return 1;
}

int decompress_data (ioport *in, ioport *out) {
    return 0;
}
