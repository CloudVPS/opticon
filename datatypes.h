#ifndef _DATATYPES_H
#define _DATATYPES_H 1

#include <stdint.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>

typedef uint64_t		meterid_t;
typedef uint64_t		metertype_t;
typedef uint32_t		status_t;

#define MTYPE_INT 		0x0000000000000000
#define MTYPE_FRAC		0x4000000000000000
#define MTYPE_STR   	0x8000000000000000

#define MMASK_TYPE		0xc000000000000000
#define MMASK_COUNT		0x0000000000000007
#define MMASK_NAME		0x3ffffffffffffff8

meterid_t 	makeid (const char *label, metertype_t type, int pos);
void	  	id2label (meterid_t id, char *into);
metertype_t	id2type (meterid_t id);

typedef struct { uint64_t msb; uint64_t lsb; } uuid_t;

typedef union {
	uint64_t		 *u64;
	double			 *frac;
	char			 *str;
	void			 *any;
} meterdata;

typedef struct meter_s {
	struct meter_s	*next;
	struct meter_s	*prev;

	meterid_t		 id;
	time_t			 lastmodified;
	int				 count;
	meterdata		 d;
} meter;

typedef struct host_s {
	struct host_s	*next;
	struct host_s	*prev;
	uuid_t			 uuid;
	status_t		 status;
	meter			*first;
	meter			*last;
} host;	

typedef struct tenant_s {
	struct tenant_s	*next;
	struct tenant_s	*prev;
	host			*first;
	host			*last;
	uuid_t			 uuid;
	char			*key;
} tenant;

typedef struct {
	tenant			*first;
	tenant			*last;
} tenantlist;

extern tenantlist TENANTS;

tenant *tenant_find (uuid_t tenantid);
tenant *tenant_create (uuid_t tenantid, const char *key);

host *host_alloc (void);
host *host_find (uuid_t tenantid, uuid_t hostid);

meter *host_get_meter (host *h, meterid_t id);
meter *host_set_meter_uint (host *h, meterid_t id, 
							unsigned int count,
							uint64_t *data);
meter *host_set_meter_frac (host *h, meterid_t id,
							unsigned int count,
							double *data);
meter *host_set_meter_str  (host *h, meterid_t id, const char *str);

uint32_t	meter_get_uint (meter *, unsigned int pos);
double		meter_get_frac (meter *, unsigned int pos);
const char *meter_get_str (meter *);

#endif
