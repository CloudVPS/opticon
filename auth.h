#ifndef _AUTH_H
#define _AUTH_H 1

#include <datatypes.h>
#include <time.h>

typedef struct aeskey_s {
    uint8_t data[32];
} aeskey;

typedef struct session_s {
    struct session_s    *next;
    struct session_s    *prev;
    uuid                 tenantid;
    uuid                 hostid;
    uint32_t             addr;
    uint32_t             sessid;
    aeskey               key;
    time_t               lastcycle;
    host                *host;
} session;

typedef struct sessionlist_s {
    session             *first;
    session             *last;
} sessionlist;

extern sessionlist SESSIONS[256];

aeskey       aeskey_create (void);

void         sessionlist_init (void);
session     *session_alloc (void);
void         session_link (session *);

session     *session_register (uuid tenantid, uuid hostid, 
                               uint32_t addrpart, uint32_t sess_id,
                               aeskey sess_key);
                               
session     *session_find (uint32_t addr, uint32_t sess_id);
void         session_expire (time_t);

#endif
