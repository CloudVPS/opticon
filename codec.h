#ifndef _CODEC_H
#define _CODEC_H 1

#include <datatypes.h>
#include <ioport.h>

/* =============================== TYPES =============================== */

typedef int (*encode_host_func)(ioport *, host *);
typedef int (*decode_host_func)(ioport *, host *);

typedef struct codec_s {
    encode_host_func    encode_host;
    decode_host_func    decode_host;
} codec;

/* ============================= FUNCTIONS ============================= */

codec       *codec_create_json (void);
codec       *codec_create_opticon (void);

int          jsoncodec_encode (ioport *, host *);
int          jsoncodec_decode (ioport *, host *);

#endif
