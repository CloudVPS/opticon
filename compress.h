#ifndef _COMPRESS_H
#define _COMPRESS_H 1

#include <datatypes.h>
#include <ioport.h>
#include <zlib.h>

/* ============================= FUNCTIONS ============================= */

int compress_data (ioport *in, ioport *out);
int decompress_data (ioport *in, ioport *out);

#endif
