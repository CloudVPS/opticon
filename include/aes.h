#ifndef _AES_H
#define _AES_H 1

#include <stdlib.h>
#include <datatypes.h>
#include <ioport.h>
#include <auth.h>

/* =============================== TYPES =============================== */

typedef struct {
    uint8_t key[32]; 
    uint8_t enckey[32]; 
    uint8_t deckey[32];
} aes256_context; 

/* ============================= FUNCTIONS ============================= */

void aes256_init(aes256_context *, uint8_t * /* key */);
void aes256_done(aes256_context *);
void aes256_encrypt_ecb(aes256_context *, uint8_t * /* plaintext */);
void aes256_decrypt_ecb(aes256_context *, uint8_t * /* cipertext */);

int ioport_encrypt (aeskey *k, ioport *in, ioport *out, time_t ts);
int ioport_decrypt (aeskey *k, ioport *in, ioport *out, time_t ts);

#endif
