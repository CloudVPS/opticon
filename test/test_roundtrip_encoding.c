#include <datatypes.h>
#include <util.h>
#include <assert.h>
#include <stdio.h>
#include <auth.h>
#include <codec.h>
#include <compress.h>
#include <aes.h>

#define FD_STDOUT 1

void hexdump_buffer (ioport *io) {
    size_t sz = ioport_read_available (io);
    size_t i;
    char *buf = ioport_get_buffer (io);
    if (! buf) return;
    
    for (i=0; i<sz; ++i) {
        printf ("%02x ", (uint8_t) buf[i]);
        if ((i&31) == 31) printf ("\n");
    }
    if (i&31) printf ("\n");
}

int main (int argc, const char *argv[]) {
    const char *stenantid = "001b71534f4b4f1cb281cc06b134f98f";
    const char *shostid = "6f943a0d-bcd9-42fa-b0c9-6ede92f9a46a";
    char tstr[16];
    time_t tnow = time (NULL);
    int i;
    uuid tenantid = mkuuid (stenantid);
    uuid hostid = mkuuid (shostid);
    meterid_t M_NET_NAME = makeid ("net/name",MTYPE_STR,0);
    meterid_t M_NET_IN_KBS = makeid ("net/in.kbs",MTYPE_INT,0);
    meterid_t M_NET_OUT_KBS = makeid ("net/out.kbs",MTYPE_INT,0);
    meterid_t M_NET_IN_PPS = makeid ("net/in.pps",MTYPE_INT,0);
    meterid_t M_NET_OUT_PPS = makeid ("net/out.pps",MTYPE_INT,0);
    meterid_t M_HOSTNAME = makeid ("hostname",MTYPE_STR,0);
    meterid_t M_UNAME = makeid ("uname",MTYPE_STR,0);

    fstring D_NET_NAME[2] = {{"eth0"},{"eth1"}};
    uint64_t D_NET_IN_KBS[2] = {100ULL,70ULL};
    uint64_t D_NET_OUT_KBS[2] = {250ULL, 90ULL};
    uint64_t D_NET_IN_PPS[2] = {150ULL,100ULL};
    uint64_t D_NET_OUT_PPS[2] = {250ULL,117ULL};
    
    meterid_t M_TOP_PID = makeid ("top/pid",MTYPE_INT,0);
    meterid_t M_TOP_USER = makeid ("top/user",MTYPE_STR,0);
    meterid_t M_TOP_PCPU = makeid ("top/pcpu",MTYPE_FRAC,0);
    meterid_t M_TOP_PMEM = makeid ("top/pmem",MTYPE_FRAC,0);
    meterid_t M_TOP_NAME = makeid ("top/name",MTYPE_STR,0);
    
    meterid_t M_MEM_SIZE = makeid ("mem/size",MTYPE_INT,0);
    meterid_t M_MEM_FREE = makeid ("mem/free",MTYPE_INT,0);
    meterid_t M_SWAP_SIZE = makeid ("swap/size",MTYPE_INT,0);
    meterid_t M_SWAP_FREE = makeid ("swap/free",MTYPE_INT,0);
    
    meterid_t M_PCPU = makeid ("pcpu",MTYPE_FRAC,0);
    meterid_t M_PIOWAIT = makeid ("piowait",MTYPE_FRAC,0);
    meterid_t M_LOADAVG = makeid ("loadavg",MTYPE_FRAC,0);
    
    uint64_t D_TOP_PID[11] = {
        27562,28074,27983,27560,28311,19082,2411,2082,6129,1862,2564
    };
    const char *D_TOP_USER[11] = {
        "www-data",
        "www-data",
        "www-data",
        "www-data",
        "www-data",
        "www-data",
        "root",
        "mysql",
        "root",
        "nobody",
        "root"
    };
    const double D_TOP_PCPU[11] = {
        16.92, 16.92, 16.88, 16.88, 16.88, 16.88, 0.03, 0.03, 0.00, 0.29, 0.03
    };
    const double D_TOP_PMEM[11] = {
        0.14,0.14,0.14,0.14,0.14,0.14,10.94,5.37,5.54,0.09,0.49
    };
    const char *D_TOP_NAME[11] = {
        "apache2",
        "apache2",
        "apache2",
        "apache2",
        "apache2",
        "apache2",
        "zarafa-server",
        "mysqld",
        "spamd",
        "n2txd",
        "zarafa-search"
    };
    
    tenant_init();
    sessionlist_init();
    
    tenant_create (tenantid, "test");
    tenant *T = tenant_find (tenantid);
    assert (T->uuid.lsb == 0x001b71534f4b4f1c);
    assert (T->uuid.msb == 0xb281cc06b134f98f);
    host *H = host_find (tenantid, hostid);
    host_begin_update (H);

    id2str (M_NET_IN_KBS, tstr);
    meter *M = host_get_meter (H, M_HOSTNAME);
    meter_setcount (M, 0);
    meter_set_str (M, 0, "srv-01.heikneuter.nl");
    
    M = host_get_meter (H, M_UNAME);
    meter_setcount (M, 0);
    meter_set_str (M, 0, "Linux srv-01.heikneuter.nl 2.6.18-prep #1 SMP");
    
    M = host_get_meter (H, M_PCPU);
    meter_setcount (M, 0);
    meter_set_frac (M, 0, 100.0);
    
    M = host_get_meter (H, M_PIOWAIT);
    meter_setcount (M, 0);
    meter_set_frac (M, 0, 3.0);
    
    M = host_get_meter (H, M_LOADAVG);
    meter_setcount (M, 0);
    meter_set_frac (M, 0, 6.03);
    
    M = host_get_meter (H, M_MEM_SIZE);
    meter_setcount (M, 0);
    meter_set_uint (M, 0, 524288);
    
    M = host_get_meter (H, M_MEM_FREE);
    meter_setcount (M, 0);
    meter_set_uint (M, 0, 424250);
    
    M = host_get_meter (H, M_SWAP_SIZE);
    meter_setcount (M, 0);
    meter_set_uint (M, 0, 1048576);
    
    M = host_get_meter (H, M_SWAP_FREE);
    meter_setcount (M, 0);
    meter_set_uint (M, 0, 122020);
    
    M = host_set_meter_str (H, M_NET_NAME, 2, D_NET_NAME);
    M = host_set_meter_uint (H, M_NET_IN_KBS, 2, D_NET_IN_KBS);
    M = host_set_meter_uint (H, M_NET_IN_PPS, 2, D_NET_IN_PPS);
    assert (meter_get_uint (M, 1) == 100ULL);
    M = host_set_meter_uint (H, M_NET_OUT_KBS, 2, D_NET_OUT_KBS);
    M = host_set_meter_uint (H, M_NET_OUT_PPS, 2, D_NET_OUT_PPS);

    M = host_get_meter (H, M_TOP_PID);
    meter_setcount (M, 11);
    for (i=0; i<11; ++i) meter_set_uint (M, i, D_TOP_PID[i]);
    
    M = host_get_meter (H, M_TOP_USER);
    meter_setcount (M, 11);
    for (i=0; i<11; ++i) meter_set_str (M, i, D_TOP_USER[i]);
    
    M = host_get_meter (H, M_TOP_PCPU);
    meter_setcount (M, 11);
    for (i=0; i<11; ++i) meter_set_frac (M, i, D_TOP_PCPU[i]);
    
    M = host_get_meter (H, M_TOP_PMEM);
    meter_setcount (M, 11);
    for (i=0; i<11; ++i) meter_set_frac (M, i, D_TOP_PMEM[i]);
    
    M = host_get_meter (H, M_TOP_NAME);
    meter_setcount (M, 11);
    for (i=0; i<11; ++i) meter_set_str (M, i, D_TOP_NAME[i]);
    
    aeskey key = aeskey_create();
    session *S = session_register (tenantid, hostid,
                                   0x0a000001, 0x31337666,
                                   key);
    
    S = session_find (0x0a000001, 0x31337666);
    assert (S != NULL);
    session_expire (time(NULL)+1);
    S = session_find (0x0a000001, 0x31337666);
    assert (S == NULL);
    
    char bouncebuf[4096];
    
    codec *C = codec_create_pkt();
    ioport *IO = ioport_create_buffer (bouncebuf, 4096);
    
    if (! codec_encode_host (C, IO, H)) {
        fprintf (stderr, "Encode failed\n");
        return 1;
    }
    
    bufferstorage *stor = (bufferstorage*) IO->storage;
    printf ("\n--> Encoded in %i bytes\n", stor->pos);
    size_t orig_size = stor->pos;
    
    char compressed[4096];
    ioport *CmpIO = ioport_create_buffer (compressed, 4096);
    compress_data (IO, CmpIO);
    printf ("--> Compressed %lu bytes\n", ioport_read_available (CmpIO));
    
    char encrypted[4096];
    ioport *CryptIO = ioport_create_buffer (encrypted, 4096);
    ioport_encrypt (&key, CmpIO, CryptIO, tnow);
    printf ("--> Encrypted %lu bytes\n", ioport_read_available (CryptIO));
    
    char decrypted[4096];
    ioport *DecrIO = ioport_create_buffer (decrypted, 4096);
    ioport_decrypt (&key, CryptIO, DecrIO, tnow);
    printf ("--> Decrypted %lu bytes\n", ioport_read_available (DecrIO));
    
    char decompressed[4096];
    ioport *DcmpIO = ioport_create_buffer (decompressed, 4096);
    decompress_data (DecrIO, DcmpIO);
    size_t bounce_size = ioport_read_available (DcmpIO);
    printf ("--> Decompressed %lu bytes\n\n", bounce_size);

    assert (bounce_size == orig_size);

    host *HH = host_alloc();
    host_begin_update (HH);
    if (! codec_decode_host (C, DcmpIO, HH)) {
        fprintf (stderr, "Decode failed\n");
        return 1;
    }
    
    ioport_close (DcmpIO);
    ioport_close (DecrIO);
    ioport_close (CryptIO);
    ioport_close (CmpIO);
    ioport_close (IO);
    codec_release (C);
    
    C = codec_create_json();
    IO = ioport_create_filewriter (stdout);
    codec_encode_host (C, IO, HH);
    ioport_close (IO);
    codec_release (C);
        
    M = host_get_meter (H, M_NET_IN_PPS);
    M = meter_next_sibling (M);
    assert (M != NULL);
    host_delete (H);
    host_delete (HH);
    return 0;
}
