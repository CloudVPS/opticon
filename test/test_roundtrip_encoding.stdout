{
  "hostname":"srv-01.heikneuter.nl",
  "uname":"Linux srv-01.heikneuter.nl 2.6.18-prep #1 SMP",
  "problems":[],
  "pcpu":100.000,
  "piowait":3.000,
  "loadavg":6.027,
  "mem":{"size":524288,"free":424250},
  "swap":{"size":1048576,"free":122020},
  "net":[
    {"name":"eth0","in.kbs":100,"in.pps":150,"out.kbs":250,"out.pps":250},
    {"name":"eth1","in.kbs":70,"in.pps":100,"out.kbs":90,"out.pps":117}
  ],
  "top":[
    {"pid":27562,"user":"www-data","pcpu":16.914,"pmem":0.137,"name":"apache2"},
    {"pid":28074,"user":"www-data","pcpu":16.914,"pmem":0.137,"name":"apache2"},
    {"pid":27983,"user":"www-data","pcpu":16.875,"pmem":0.137,"name":"apache2"},
    {"pid":27560,"user":"www-data","pcpu":16.875,"pmem":0.137,"name":"apache2"},
    {"pid":28311,"user":"www-data","pcpu":16.875,"pmem":0.137,"name":"apache2"},
    {"pid":19082,"user":"www-data","pcpu":16.875,"pmem":0.137,"name":"apache2"},
    {"pid":2411,"user":"root","pcpu":0.027,"pmem":10.934,"name":"zarafa-server"},
    {"pid":2082,"user":"mysql","pcpu":0.027,"pmem":5.367,"name":"mysqld"},
    {"pid":6129,"user":"root","pcpu":0.000,"pmem":5.535,"name":"spamd"},
    {"pid":1862,"user":"nobody","pcpu":0.285,"pmem":0.086,"name":"n2txd"},
    {"pid":2564,"user":"root","pcpu":0.027,"pmem":0.484,"name":"zarafa-search"}
  ]
}
